Sources/FoundationEssentials/FileManager/FileOperations.swift

```diff
@@ -255,24 +300,94 @@ struct NSFileManagerMoveOptions: ExpressibleByArrayLiteral {
 #endif
 
 private protocol LinkOrCopyDelegate {
-    func shouldPerformOnItemAtPath(_ path: String, to dstPtr: UnsafePointer<CChar>) -> Bool
-    func throwIfNecessary(_ errno: Int32, _ srcPtr: UnsafePointer<CChar>, _ dstPtr: UnsafePointer<CChar>) throws
-    func throwIfNecessary(_ error: Error, _ srcString: String, _ dstString: String) throws
+#if os(Windows)
+    typealias ErrorType = DWORD
+#else
+    typealias ErrorType = Int32
+#endif
+
+    func shouldPerformOnItemAtPath(_ path: String, to destination: String) -> Bool
+    func throwIfNecessary(_ errno: ErrorType, _ source: String, _ destination: String) throws
     var extraCopyFileFlags: Int32 { get }
     var copyData: Bool { get }
 }
 
 private extension LinkOrCopyDelegate {
     var extraCopyFileFlags: Int32 { 0 }
-    func throwIfNecessary(_ error: Error, _ srcPtr: UnsafePointer<CChar>, _ dstPtr: UnsafePointer<CChar>) throws {
-        let srcString = String(cString: srcPtr)
-        let dstString = String(cString: dstPtr)
-        try throwIfNecessary(error, srcString, dstString)
-    }
 }
 
 enum _FileOperations {
     // MARK: removefile
+
+#if os(Windows)
+    static func removeFile(_ path: String, with filemanager: FileManager?) throws {
+        guard filemanager?._shouldRemoveItemAtPath(path) ?? true else { return }
+        try path.withNTPathRepresentation {
+            var faAttributes: WIN32_FILE_ATTRIBUTE_DATA = .init()
+            guard GetFileAttributesExW($0, GetFileExInfoStandard, &faAttributes) else {
+                let error = CocoaError.removeFileError(GetLastError(), path)
+                guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: path) ?? false) else {
+                    throw error
+                }
+                return
+            }
+            if faAttributes.dwFileAttributes & FILE_ATTRIBUTE_READONLY == FILE_ATTRIBUTE_READONLY {
+                guard SetFileAttributesW($0, faAttributes.dwFileAttributes & ~FILE_ATTRIBUTE_READONLY) else {
+                    throw CocoaError.removeFileError(GetLastError(), path)
+                }
+            }
+            if faAttributes.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == 0 || faAttributes.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT == FILE_ATTRIBUTE_REPARSE_POINT {
+                if faAttributes.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == FILE_ATTRIBUTE_DIRECTORY {
+                    guard RemoveDirectoryW($0) else {
+                        throw CocoaError.removeFileError(GetLastError(), path)
+                    }
+                } else {
+                    guard DeleteFileW($0) else {
+                        throw CocoaError.removeFileError(GetLastError(), path)
+                    }
+                }
+            }
+        }
+
+        var stack = [path]
+        while let directory = stack.popLast() {
+            guard filemanager?._shouldRemoveItemAtPath(directory) ?? true else { continue }
```

Comment:
I think this will duplicate the call at line 324 above, should we try to avoid asking twice?

Reply:
Yes, the question will be asked twice for the first entry in the case that we hit a non-symlinked directory. We can sink the first check to before the removal attempt.